File: 1855-maximum-distance-between-a-pair-of-values/solution.py
class Solution:
    def maxDistance(self, nums1: list[int], nums2: list[int]) -> int:
        """
        Calculates the maximum distance j - i using a two-pointer technique.
        """
        i, j = 0, 0
        max_dist = 0
        n1, n2 = len(nums1), len(nums2)

        while i < n1 and j < n2:
            # Check the validity of the current pair (i, j)
            if nums1[i] <= nums2[j]:
                # Update max_dist if the current distance is larger
                # Python's max handles the case where j < i (negative result)
                if j >= i:
                    max_dist = max(max_dist, j - i)
                
                # Try expanding j to the right for a larger distance
                j += 1
            else:
                # nums1[i] is too large for the current nums2[j].
                # Move i to a smaller value (arrays are non-increasing).
                i += 1
                
        return max_dist

File: 1855-maximum-distance-between-a-pair-of-values/solution.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    /**
     * Finds the maximum distance j - i such that i <= j and nums1[i] <= nums2[j].
     * Uses a two-pointer approach leveraging the non-increasing property of both arrays.
     */
    int maxDistance(vector<int>& nums1, vector<int>& nums2) {
        int i = 0;
        int j = 0;
        int maxDist = 0;
        int n1 = nums1.size();
        int n2 = nums2.size();

        while (i < n1 && j < n2) {
            // Check if the condition is satisfied
            if (nums1[i] <= nums2[j]) {
                // Since i <= j is required, and we start both at 0, 
                // we only calculate distance if i <= j.
                // However, the logic below naturally handles i <= j 
                // because if i > j, nums1[i] <= nums2[j] might be true,
                // but j - i would be negative, which won't update maxDist.
                maxDist = max(maxDist, j - i);
                
                // Try to increase j to maximize the distance
                j++;
            } else {
                // If nums1[i] is too large, move i to a smaller value
                i++;
                // Optimization: Ensure i doesn't lag too far behind j 
                // to maintain i <= j logic, though max(0, j-i) handles this.
            }
        }

        return maxDist;
    }
};